File: java/examples/src/main/java/Basic.java
import com.visualthreat.api.API;
import com.visualthreat.api.VTCloud;
import com.visualthreat.api.data.*;
import com.visualthreat.api.v1.Token;

import java.util.Collections;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

public class Basic {
  public static void main(String[] args) {
    // create API object
    final API api = API.get();

    // use test api key and secret
    final String key = "700fb92b-1505-4a39-9b08-927cbe0257d6";
    final String secret = "47ac4af9-ed2a-4ed0-8d59-d431162a3032";

    // authenticate
    final Token token = api.authenticate(key, secret);
    Device device = null;
    for (final Device d : api.getConnectedDevices(token)) {
      // get available devices, Device.TEST_DEVICE is always available
      // compare with Device.getName if you want to use specific one
      if (d.isAvailable()) {
        device = d;
        break;
      }
    }

    // connect to found device and get VTCloud object
    final VTCloud cloud = api.connectToDevice(device, token);

    // create CAN frame, with id 0x700 and data
    final CANFrame canFrame = new CANFrame(0x700, new byte[]{(byte) 0x01});
    // create request to send the can frame
    // and collect all responses from CAN bus for 300 ms
    final Request request = new Request(canFrame, 300);
    // send it, without any response filter
    System.out.println("Sending CAN request...");
    final Iterator<Response> responses = cloud.sendCANFrames(
        Collections.singletonList(request), CANResponseFilter.NONE);

    final Set<String> receivedIDs = new HashSet<>();
    // iterate through all request-responses pairs
    while (responses.hasNext()) {
      final Response response = responses.next();
      final Iterator<CANFrame> responseFrames = response.getResponses();
      // iterate through response frames (you can use Response.getRequest to see request)
      while (responseFrames.hasNext()) {
        // get all unique IDs from received can frames
        receivedIDs.add("0x" + Integer.toHexString(responseFrames.next().getId()));
      }
    }
    cloud.close();

    // print the result
    System.out.println(String.format("%d different ids received: %s",
        receivedIDs.size(), receivedIDs.toString()));
  }
}

File: python/examples/basic.py
from sdk.API import API
from sdk.data.CANFrame import CANFrame
from sdk.data.CANResponseFilter import CANResponseFilter
from sdk.data.Query import Query
from sdk.data.Request import Request

if __name__ == "__main__":
    # use test api key and secret
    key = "0dee4f2e-4b05-445d-b8ea-f8fe6b4b772c"
    secret = "1bcdd94c-a7ed-4e24-b601-6e8753ca3721"
    # create api connection
    api = API()
    # get authentication token
    token = api.authenticate(key, secret)

    # get available devices, Device.TEST_DEVICE is always available
    devices = api.get_connected_devices(token)
    # find available device
    device = None
    for dev in devices:
        if dev.is_available_now():
            device = dev
            print("Selected car: " + device.name)
            break

    # connect to it
    cloud = api.connect_to_device(token, device)

    # create CAN frame, with id 0x700 and data
    can_frame = CANFrame(0x700, bytearray([0x1]))
    # create request to send the can frame
    # and collect all responses from CAN bus for 300 ms
    request = Request(can_frame, 300)
    try:
        can_query = Query([request], CANResponseFilter.NONE())
        responses = cloud.send_can_query(can_query)
        for response in responses:
            frame_ids = {hex(x.frame_id) for x in response.iterator()}
            print("Request ID: " + hex(response.request.frame_id))
            print("Unique ids: " + str(len(frame_ids)))
            print("IDs: " + str(frame_ids))
    finally:
        cloud.close()
